File: python/permutations.py
def permutations(lst):
    return permutations_r([[]], lst)


def permutations_r(lst, remaining):
    if len(remaining) == 0:
        return lst
    remaining = remaining.copy()
    result = []
    for i, v in enumerate(remaining):
        cp = remaining.copy()
        cp.pop(i)
        result += permutations_r([l + [v] for l in lst], cp)
    return result

File: src/permutations.go
package solver24

func Permutations[T any](inp []T) [][]T {
	return permutationsRecursive([][]T{{}}, inp)
}

func permutationsRecursive[T any](gen [][]T, inp []T) [][]T {
	if len(inp) == 0 {
		return gen
	}

	result := [][]T{}
	for i, v := range inp {
		inpTemp := make([]T, len(inp))
		copy(inpTemp, inp)
		inpTemp = append(inpTemp[:i], inpTemp[i+1:]...)

		gen2 := make([][]T, len(gen))
		copy(gen2, gen)
		for i2, v2 := range gen2 {
			gen2[i2] = append(v2, v)
		}
		result = append(result, permutationsRecursive(gen2, inpTemp)...)
	}
	return result
}
